{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "147e99a0",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Filter-out-interactions-based-on-taxonomy\" data-toc-modified-id=\"Filter-out-interactions-based-on-taxonomy-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Filter out interactions based on taxonomy</a></span></li><li><span><a href=\"#Filter-out-microbes-based-on-taxonomy\" data-toc-modified-id=\"Filter-out-microbes-based-on-taxonomy-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Filter out microbes based on taxonomy</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c00962e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This block is just for importing the necessary libraries.  \n",
    "import os\n",
    "from collections import defaultdict\n",
    "# Numerical libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import arviz as az\n",
    "from scipy.spatial.distance import euclidean\n",
    "# Plotting libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2, venn3\n",
    "# custom utility and plotting functions\n",
    "from util import (extract_differentials, select_features, \n",
    "                  get_genomic_data, collapse_transcripts, \n",
    "                  aggregate_pathways,\n",
    "                  ranking, btest, log_pvalue, read_kegg_dict,\n",
    "                  ilr_transform_differentials,\n",
    "                  rename_clades, create_projection,\n",
    "                  match_all_differentials)\n",
    "from plot import (rankplot, networkplot, vectorplot)\n",
    "import random\n",
    "\n",
    "# directory paths\n",
    "dan_directory = '../sfari/data/sra_shotgun/Dan2020'\n",
    "averina_directory = '../sfari/data/sra_shotgun/Averina2020'\n",
    "wang_directory = '../sfari/data/sra_shotgun/Wang2021'\n",
    "taxa_directory = '~/databases/wol/taxonomy'\n",
    "taxa_directory = '~/databases/wol/taxonomy'\n",
    "gtdb_directory = '~/databases/wol/taxonomy/gtdb'\n",
    "results_dir = '../results'\n",
    "mmvec_edge_dir = f'{results_dir}/mmvec/Network'\n",
    "kegg_dir = f'{results_dir}/kegg'\n",
    "hsa_dir = f'{results_dir}/hsa_kegg'\n",
    "np.random.seed(0)\n",
    "random.seed(0)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "raw",
   "id": "da0bf9d7",
   "metadata": {},
   "source": [
    "# Load taxonomies\n",
    "taxids = pd.read_table(f'{taxa_directory}/taxid.map', header=None, dtype=str)\n",
    "ranks = pd.read_table(f'{taxa_directory}/ranks.tsv')\n",
    "taxonomy = pd.merge(ranks, taxids, left_on='genome', right_on=0)\n",
    "taxonomy = taxonomy.set_index(1)\n",
    "del taxonomy[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87528240",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load taxonomies\n",
    "taxids = pd.read_table(f'{taxa_directory}/taxid.map', header=None, dtype=str)\n",
    "ranks = pd.read_table(f'{gtdb_directory}/ranks.tsv')\n",
    "taxonomy = pd.merge(ranks, taxids, left_on='gid', right_on=0)\n",
    "taxonomy = taxonomy.set_index(1)\n",
    "del taxonomy[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d56570c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_slash(x):\n",
    "    #k = x['kingdom']\n",
    "    p = x['phylum']\n",
    "    c = x['class']\n",
    "    o = x['order']\n",
    "    f = x['family']\n",
    "    g = x['genus']\n",
    "    s = x['species']\n",
    "    return '/'.join(list(map(str, [p, c, o, f, g, s])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24e3331d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wol_taxa = list(taxonomy.apply(concat_slash, axis=1).values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3d1403a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gpd_metadata = pd.read_table('../results/mmvec/GPD_metadata.tsv')\n",
    "gpd_metadata = gpd_metadata.dropna(subset=['Host_range_isolates'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9f5d6088",
   "metadata": {},
   "outputs": [],
   "source": [
    "lookup = {\n",
    " 'Actinobacteriota' : 'Actinobacteria',\n",
    " 'Bacteroidota' : 'Bacteroidetes',\n",
    " 'Desulfobacterota': 'Deltaproteobacteria',\n",
    " 'Synergistota': 'Synergistetes',\n",
    " 'Campylobacterota': 'Epsilonproteobacteria',\n",
    " 'Fusobacteriota': 'Fusobacteria'}\n",
    "    \n",
    "def fix_f(x):\n",
    "    y = x.split('/')[0]\n",
    "    ## uncomment if you want to fix phylum level naming\n",
    "    #if y in lookup.keys():\n",
    "    #    z = lookup[y]\n",
    "    #    return x.replace(y, z)\n",
    "    return x\n",
    "        \n",
    "def all_taxa(x):\n",
    "    if pd.isnull(x):\n",
    "        return []\n",
    "    else:\n",
    "        taxa = str(x).split(',')\n",
    "        return list(set(list(map(fix_f, taxa))))\n",
    "    \n",
    "gpd_taxa = list(gpd_metadata['Host_range_taxon'].apply(all_taxa).values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f24dd2",
   "metadata": {},
   "source": [
    "# Filter out interactions based on taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "029562e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_gpd_taxa = sum(gpd_taxa, [])\n",
    "all_gpd_taxa = set(all_gpd_taxa)\n",
    "common_taxa = set(wol_taxa) & set(all_gpd_taxa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2bb35fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def contains_f(x):\n",
    "    return len(set(x) & common_taxa) > 0\n",
    "\n",
    "idx = list(map(contains_f, gpd_taxa))\n",
    "\n",
    "gpd_subset = gpd_metadata.loc[idx]\n",
    "viruses = gpd_metadata.loc[idx, 'GPD_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "836a0f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "averina_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Averina.txt')\n",
    "averina_edges = averina_edges.loc[averina_edges.cond_logit > 1]\n",
    "\n",
    "dan_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Dan.txt')\n",
    "dan_edges = dan_edges.loc[dan_edges.cond_logit > 1]\n",
    "\n",
    "pths_edges = pd.read_table(f'{mmvec_edge_dir}/edges_PTHS.txt')\n",
    "pths_edges = pths_edges.loc[pths_edges.cond_logit > 1]\n",
    "\n",
    "wang_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Wang.txt')\n",
    "wang_edges = wang_edges.loc[wang_edges.cond_logit > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "729c9fa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23896, 3) (20584, 3) (34100, 3)\n"
     ]
    }
   ],
   "source": [
    "print(dan_edges.shape, wang_edges.shape, averina_edges.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "12e95ef2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78580"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dan_edges.shape[0] + wang_edges.shape[0] + averina_edges.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e1f12b",
   "metadata": {},
   "source": [
    "# Filter out microbes based on taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f231baff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def contains_taxonomy(x):\n",
    "    x = str(x)\n",
    "    if x in taxonomy.index:\n",
    "        tstr = concat_slash(taxonomy.loc[x])\n",
    "        return tstr in common_taxa\n",
    "    return False\n",
    "wang_edges = wang_edges.loc[wang_edges.microbe.apply(contains_taxonomy)]\n",
    "dan_edges = dan_edges.loc[dan_edges.microbe.apply(contains_taxonomy)]\n",
    "averina_edges = averina_edges.loc[averina_edges.microbe.apply(contains_taxonomy)]\n",
    "pths_edges = pths_edges.loc[pths_edges.microbe.apply(contains_taxonomy)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9ed1d6e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10626, 3) (10574, 3) (10076, 3)\n"
     ]
    }
   ],
   "source": [
    "print(dan_edges.shape, wang_edges.shape, averina_edges.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aa2e4c11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31276"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dan_edges.shape[0] + wang_edges.shape[0] + averina_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6dd550ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "wang_edges['dataset'] = 'Wang'\n",
    "dan_edges['dataset'] = 'Dan'\n",
    "averina_edges['dataset'] = 'Averina'\n",
    "# pths_edges['dataset'] = 'PTHS'\n",
    "all_edges = pd.concat((wang_edges, dan_edges, averina_edges), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1f030ceb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-16-097f4499dd13>:1: DeprecationWarning: `np.str` is a deprecated alias for the builtin `str`. To silence this warning, use `str` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.str_` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  all_edges['microbe'] = all_edges['microbe'].astype(np.str)\n"
     ]
    }
   ],
   "source": [
    "all_edges['microbe'] = all_edges['microbe'].astype(np.str)\n",
    "# merge microbial metadata\n",
    "all_edges = pd.merge(all_edges, gpd_metadata, left_on='virus', right_on='GPD_id')\n",
    "# merge GPD metadata\n",
    "all_edges = pd.merge(all_edges, taxonomy, left_on='microbe', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e9f45cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_edges.to_csv('../results/interaction_metadata.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8f9639d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>microbe</th>\n",
       "      <th>virus</th>\n",
       "      <th>cond_logit</th>\n",
       "      <th>dataset</th>\n",
       "      <th>GPD_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>GPD_VC</th>\n",
       "      <th>Size</th>\n",
       "      <th>Predicted_phage_taxon</th>\n",
       "      <th>Host_range_isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>checkV_termini</th>\n",
       "      <th>Novel</th>\n",
       "      <th>gid</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1262981</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>4.954891</td>\n",
       "      <td>Wang</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>37815</td>\n",
       "      <td>32601</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>GCA_003464755,GCF_000164195,12718_7#19,20298_3...</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G000435955</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Erysipelotrichales</td>\n",
       "      <td>Erysipelotrichaceae</td>\n",
       "      <td>Eubacterium_D</td>\n",
       "      <td>Eubacterium_D innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1262981</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>4.954891</td>\n",
       "      <td>Wang</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>37815</td>\n",
       "      <td>32601</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>GCA_003464755,GCF_000164195,12718_7#19,20298_3...</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G000435955</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Erysipelotrichales</td>\n",
       "      <td>Erysipelotrichaceae</td>\n",
       "      <td>Eubacterium_D</td>\n",
       "      <td>Eubacterium_D innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1262981</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>4.954647</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18797</td>\n",
       "      <td>40415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000011045</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G000435955</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Erysipelotrichales</td>\n",
       "      <td>Erysipelotrichaceae</td>\n",
       "      <td>Eubacterium_D</td>\n",
       "      <td>Eubacterium_D innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>1262981</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>4.954647</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18797</td>\n",
       "      <td>40415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000011045</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G000435955</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Erysipelotrichales</td>\n",
       "      <td>Erysipelotrichaceae</td>\n",
       "      <td>Eubacterium_D</td>\n",
       "      <td>Eubacterium_D innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>1262981</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>4.915556</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>41768</td>\n",
       "      <td>39592</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003602765</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G000435955</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Erysipelotrichales</td>\n",
       "      <td>Erysipelotrichaceae</td>\n",
       "      <td>Eubacterium_D</td>\n",
       "      <td>Eubacterium_D innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6007</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>1.052579</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>3999</td>\n",
       "      <td>59704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003482185</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6041</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_568287</td>\n",
       "      <td>1.173307</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_568287</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>908</td>\n",
       "      <td>119423</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12718_7#33,GCA_003466785</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6066</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_568287</td>\n",
       "      <td>1.173307</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_568287</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>908</td>\n",
       "      <td>119423</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12718_7#33,GCA_003466785</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6189</th>\n",
       "      <td>1871003</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>1.122697</td>\n",
       "      <td>Averina</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18307</td>\n",
       "      <td>23261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21673_4#3</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6212</th>\n",
       "      <td>1871003</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>1.122697</td>\n",
       "      <td>Averina</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18307</td>\n",
       "      <td>23261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21673_4#3</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6530 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      microbe        virus  cond_logit  dataset       GPD_id      Source  \\\n",
       "0     1262981  uvig_108002    4.954891     Wang  uvig_108002  Metagenome   \n",
       "22    1262981  uvig_108002    4.954891     Wang  uvig_108002  Metagenome   \n",
       "44    1262981    ivig_2376    4.954647     Wang    ivig_2376     Isolate   \n",
       "66    1262981    ivig_2376    4.954647     Wang    ivig_2376     Isolate   \n",
       "88    1262981    ivig_2256    4.915556     Wang    ivig_2256     Isolate   \n",
       "...       ...          ...         ...      ...          ...         ...   \n",
       "6007  1871003  uvig_492171    1.052579  Averina  uvig_492171  Metagenome   \n",
       "6041  1871003  uvig_568287    1.173307  Averina  uvig_568287  Metagenome   \n",
       "6066  1871003  uvig_568287    1.173307  Averina  uvig_568287  Metagenome   \n",
       "6189  1871003    ivig_4295    1.122697  Averina    ivig_4295     Isolate   \n",
       "6212  1871003    ivig_4295    1.122697  Averina    ivig_4295     Isolate   \n",
       "\n",
       "      GPD_VC    Size Predicted_phage_taxon  \\\n",
       "0      37815   32601            Myoviridae   \n",
       "22     37815   32601            Myoviridae   \n",
       "44     18797   40415                   NaN   \n",
       "66     18797   40415                   NaN   \n",
       "88     41768   39592                   NaN   \n",
       "...      ...     ...                   ...   \n",
       "6007    3999   59704                   NaN   \n",
       "6041     908  119423                   NaN   \n",
       "6066     908  119423                   NaN   \n",
       "6189   18307   23261                   NaN   \n",
       "6212   18307   23261                   NaN   \n",
       "\n",
       "                                    Host_range_isolates  ... checkV_termini  \\\n",
       "0     GCA_003464755,GCF_000164195,12718_7#19,20298_3...  ...             No   \n",
       "22    GCA_003464755,GCF_000164195,12718_7#19,20298_3...  ...             No   \n",
       "44                                        GCF_000011045  ...             No   \n",
       "66                                        GCF_000011045  ...             No   \n",
       "88                                        GCA_003602765  ...             No   \n",
       "...                                                 ...  ...            ...   \n",
       "6007                                      GCA_003482185  ...             No   \n",
       "6041                           12718_7#33,GCA_003466785  ...             No   \n",
       "6066                           12718_7#33,GCA_003466785  ...             No   \n",
       "6189                                          21673_4#3  ...             No   \n",
       "6212                                          21673_4#3  ...             No   \n",
       "\n",
       "     Novel         gid    domain        phylum        class  \\\n",
       "0      Yes  G000435955  Bacteria    Firmicutes      Bacilli   \n",
       "22     Yes  G000435955  Bacteria    Firmicutes      Bacilli   \n",
       "44     Yes  G000435955  Bacteria    Firmicutes      Bacilli   \n",
       "66     Yes  G000435955  Bacteria    Firmicutes      Bacilli   \n",
       "88     Yes  G000435955  Bacteria    Firmicutes      Bacilli   \n",
       "...    ...         ...       ...           ...          ...   \n",
       "6007   Yes  G900104605  Bacteria  Bacteroidota  Bacteroidia   \n",
       "6041    No  G900104605  Bacteria  Bacteroidota  Bacteroidia   \n",
       "6066    No  G900104605  Bacteria  Bacteroidota  Bacteroidia   \n",
       "6189    No  G900104605  Bacteria  Bacteroidota  Bacteroidia   \n",
       "6212    No  G900104605  Bacteria  Bacteroidota  Bacteroidia   \n",
       "\n",
       "                   order               family          genus  \\\n",
       "0     Erysipelotrichales  Erysipelotrichaceae  Eubacterium_D   \n",
       "22    Erysipelotrichales  Erysipelotrichaceae  Eubacterium_D   \n",
       "44    Erysipelotrichales  Erysipelotrichaceae  Eubacterium_D   \n",
       "66    Erysipelotrichales  Erysipelotrichaceae  Eubacterium_D   \n",
       "88    Erysipelotrichales  Erysipelotrichaceae  Eubacterium_D   \n",
       "...                  ...                  ...            ...   \n",
       "6007       Bacteroidales        Rikenellaceae  Tidjanibacter   \n",
       "6041       Bacteroidales        Rikenellaceae  Tidjanibacter   \n",
       "6066       Bacteroidales        Rikenellaceae  Tidjanibacter   \n",
       "6189       Bacteroidales        Rikenellaceae  Tidjanibacter   \n",
       "6212       Bacteroidales        Rikenellaceae  Tidjanibacter   \n",
       "\n",
       "                         species  \n",
       "0         Eubacterium_D innocuum  \n",
       "22        Eubacterium_D innocuum  \n",
       "44        Eubacterium_D innocuum  \n",
       "66        Eubacterium_D innocuum  \n",
       "88        Eubacterium_D innocuum  \n",
       "...                          ...  \n",
       "6007  Tidjanibacter massiliensis  \n",
       "6041  Tidjanibacter massiliensis  \n",
       "6066  Tidjanibacter massiliensis  \n",
       "6189  Tidjanibacter massiliensis  \n",
       "6212  Tidjanibacter massiliensis  \n",
       "\n",
       "[6530 rows x 29 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a01f1749",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>398511</th>\n",
       "      <td>G000005825</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae_D</td>\n",
       "      <td>Bacillus_S</td>\n",
       "      <td>Bacillus_S pseudofirmus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456320</th>\n",
       "      <td>G000006175</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanococcaceae</td>\n",
       "      <td>Methanococcus</td>\n",
       "      <td>Methanococcus voltae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306537</th>\n",
       "      <td>G000006605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Corynebacterium</td>\n",
       "      <td>Corynebacterium jeikeium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160492</th>\n",
       "      <td>G000006725</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Xylella</td>\n",
       "      <td>Xylella fastidiosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243277</th>\n",
       "      <td>G000006745</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484498</th>\n",
       "      <td>G900156675</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Saccharospirillaceae</td>\n",
       "      <td>Oleibacter</td>\n",
       "      <td>Oleibacter marinus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477680</th>\n",
       "      <td>G900156765</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Filimonas</td>\n",
       "      <td>Filimonas lacunae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926284</th>\n",
       "      <td>G900156885</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus_H</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926277</th>\n",
       "      <td>G900156915</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Amphibacillaceae</td>\n",
       "      <td>Sediminibacillus_A</td>\n",
       "      <td>Sediminibacillus_A massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255658</th>\n",
       "      <td>G900163845</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Propionibacteriales</td>\n",
       "      <td>Propionibacteriaceae</td>\n",
       "      <td>Luteococcus</td>\n",
       "      <td>Luteococcus japonicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9732 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gid    domain            phylum                class  \\\n",
       "1                                                                      \n",
       "398511   G000005825  Bacteria        Firmicutes              Bacilli   \n",
       "456320   G000006175   Archaea     Euryarchaeota         Methanococci   \n",
       "306537   G000006605  Bacteria  Actinobacteriota       Actinobacteria   \n",
       "160492   G000006725  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "243277   G000006745  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "...             ...       ...               ...                  ...   \n",
       "484498   G900156675  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "477680   G900156765  Bacteria      Bacteroidota          Bacteroidia   \n",
       "1926284  G900156885  Bacteria        Firmicutes              Bacilli   \n",
       "1926277  G900156915  Bacteria        Firmicutes              Bacilli   \n",
       "1255658  G900163845  Bacteria  Actinobacteriota       Actinobacteria   \n",
       "\n",
       "                       order                family               genus  \\\n",
       "1                                                                        \n",
       "398511            Bacillales         Bacillaceae_D          Bacillus_S   \n",
       "456320       Methanococcales      Methanococcaceae       Methanococcus   \n",
       "306537     Corynebacteriales    Corynebacteriaceae     Corynebacterium   \n",
       "160492       Xanthomonadales      Xanthomonadaceae             Xylella   \n",
       "243277      Enterobacterales          Vibrionaceae              Vibrio   \n",
       "...                      ...                   ...                 ...   \n",
       "484498       Pseudomonadales  Saccharospirillaceae          Oleibacter   \n",
       "477680       Chitinophagales      Chitinophagaceae           Filimonas   \n",
       "1926284      Lactobacillales      Lactobacillaceae     Lactobacillus_H   \n",
       "1926277           Bacillales      Amphibacillaceae  Sediminibacillus_A   \n",
       "1255658  Propionibacteriales  Propionibacteriaceae         Luteococcus   \n",
       "\n",
       "                                 species  \n",
       "1                                         \n",
       "398511           Bacillus_S pseudofirmus  \n",
       "456320              Methanococcus voltae  \n",
       "306537          Corynebacterium jeikeium  \n",
       "160492                Xylella fastidiosa  \n",
       "243277                   Vibrio cholerae  \n",
       "...                                  ...  \n",
       "484498                Oleibacter marinus  \n",
       "477680                 Filimonas lacunae  \n",
       "1926284                              NaN  \n",
       "1926277  Sediminibacillus_A massiliensis  \n",
       "1255658            Luteococcus japonicus  \n",
       "\n",
       "[9732 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f39691",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
