{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "147e99a0",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Filter-out-interactions-based-on-taxonomy\" data-toc-modified-id=\"Filter-out-interactions-based-on-taxonomy-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Filter out interactions based on taxonomy</a></span></li><li><span><a href=\"#Filter-out-microbes-based-on-taxonomy\" data-toc-modified-id=\"Filter-out-microbes-based-on-taxonomy-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Filter out microbes based on taxonomy</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c00962e7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:51.557851Z",
     "iopub.status.busy": "2021-10-28T03:58:51.557546Z",
     "iopub.status.idle": "2021-10-28T03:58:53.371113Z",
     "shell.execute_reply": "2021-10-28T03:58:53.371371Z"
    }
   },
   "outputs": [],
   "source": [
    "# This block is just for importing the necessary libraries.  \n",
    "import os\n",
    "from collections import defaultdict\n",
    "# Numerical libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import arviz as az\n",
    "from scipy.spatial.distance import euclidean\n",
    "# Plotting libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2, venn3\n",
    "# custom utility and plotting functions\n",
    "from util import (extract_differentials, select_features, \n",
    "                  get_genomic_data, collapse_transcripts, \n",
    "                  aggregate_pathways,\n",
    "                  ranking, btest, log_pvalue, read_kegg_dict,\n",
    "                  ilr_transform_differentials,\n",
    "                  rename_clades, create_projection,\n",
    "                  match_all_differentials)\n",
    "from plot import (rankplot, networkplot, vectorplot)\n",
    "import random\n",
    "\n",
    "# directory paths\n",
    "dan_directory = '../sfari/data/sra_shotgun/Dan2020'\n",
    "averina_directory = '../sfari/data/sra_shotgun/Averina2020'\n",
    "wang_directory = '../sfari/data/sra_shotgun/Wang2021'\n",
    "taxa_directory = '~/databases/wol/taxonomy'\n",
    "\n",
    "results_dir = '../results'\n",
    "mmvec_edge_dir = f'{results_dir}/mmvec/Network'\n",
    "kegg_dir = f'{results_dir}/kegg'\n",
    "hsa_dir = f'{results_dir}/hsa_kegg'\n",
    "np.random.seed(0)\n",
    "random.seed(0)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "251e6057",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:53.374107Z",
     "iopub.status.busy": "2021-10-28T03:58:53.373812Z",
     "iopub.status.idle": "2021-10-28T03:58:53.447786Z",
     "shell.execute_reply": "2021-10-28T03:58:53.447481Z"
    }
   },
   "outputs": [],
   "source": [
    "# Load taxonomies\n",
    "taxids = pd.read_table(f'{taxa_directory}/taxid.map', header=None, dtype=str)\n",
    "ranks = pd.read_table(f'{taxa_directory}/ranks.tsv')\n",
    "taxonomy = pd.merge(ranks, taxids, left_on='genome', right_on=0)\n",
    "taxonomy = taxonomy.set_index(1)\n",
    "del taxonomy[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87528240",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:53.451437Z",
     "iopub.status.busy": "2021-10-28T03:58:53.451151Z",
     "iopub.status.idle": "2021-10-28T03:58:53.475929Z",
     "shell.execute_reply": "2021-10-28T03:58:53.475640Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>398511</th>\n",
       "      <td>G000005825</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillus pseudofirmus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456320</th>\n",
       "      <td>G000006175</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanococcaceae</td>\n",
       "      <td>Methanococcus</td>\n",
       "      <td>Methanococcus voltae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306537</th>\n",
       "      <td>G000006605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Corynebacterium</td>\n",
       "      <td>Corynebacterium jeikeium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160492</th>\n",
       "      <td>G000006725</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Xylella</td>\n",
       "      <td>Xylella fastidiosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243277</th>\n",
       "      <td>G000006745</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484498</th>\n",
       "      <td>G900156675</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillaceae</td>\n",
       "      <td>Oleibacter</td>\n",
       "      <td>Oleibacter marinus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477680</th>\n",
       "      <td>G900156765</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Chitinophagia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Filimonas</td>\n",
       "      <td>Filimonas lacunae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926284</th>\n",
       "      <td>G900156885</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillus sp. Marseille-P3519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926277</th>\n",
       "      <td>G900156915</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Aquibacillus</td>\n",
       "      <td>Aquibacillus sp. Marseille-P3518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255658</th>\n",
       "      <td>G900163845</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Propionibacteriales</td>\n",
       "      <td>Propionibacteriaceae</td>\n",
       "      <td>Luteococcus</td>\n",
       "      <td>Luteococcus japonicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10575 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             genome   kingdom          phylum                class  \\\n",
       "1                                                                    \n",
       "398511   G000005825  Bacteria      Firmicutes              Bacilli   \n",
       "456320   G000006175   Archaea   Euryarchaeota         Methanococci   \n",
       "306537   G000006605  Bacteria  Actinobacteria       Actinobacteria   \n",
       "160492   G000006725  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "243277   G000006745  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "...             ...       ...             ...                  ...   \n",
       "484498   G900156675  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "477680   G900156765  Bacteria   Bacteroidetes        Chitinophagia   \n",
       "1926284  G900156885  Bacteria      Firmicutes              Bacilli   \n",
       "1926277  G900156915  Bacteria      Firmicutes              Bacilli   \n",
       "1255658  G900163845  Bacteria  Actinobacteria       Actinobacteria   \n",
       "\n",
       "                       order                family            genus  \\\n",
       "1                                                                     \n",
       "398511            Bacillales           Bacillaceae         Bacillus   \n",
       "456320       Methanococcales      Methanococcaceae    Methanococcus   \n",
       "306537     Corynebacteriales    Corynebacteriaceae  Corynebacterium   \n",
       "160492       Xanthomonadales      Xanthomonadaceae          Xylella   \n",
       "243277           Vibrionales          Vibrionaceae           Vibrio   \n",
       "...                      ...                   ...              ...   \n",
       "484498     Oceanospirillales    Oceanospirillaceae       Oleibacter   \n",
       "477680       Chitinophagales      Chitinophagaceae        Filimonas   \n",
       "1926284      Lactobacillales      Lactobacillaceae    Lactobacillus   \n",
       "1926277           Bacillales           Bacillaceae     Aquibacillus   \n",
       "1255658  Propionibacteriales  Propionibacteriaceae      Luteococcus   \n",
       "\n",
       "                                   species  \n",
       "1                                           \n",
       "398511               Bacillus pseudofirmus  \n",
       "456320                Methanococcus voltae  \n",
       "306537            Corynebacterium jeikeium  \n",
       "160492                  Xylella fastidiosa  \n",
       "243277                     Vibrio cholerae  \n",
       "...                                    ...  \n",
       "484498                  Oleibacter marinus  \n",
       "477680                   Filimonas lacunae  \n",
       "1926284  Lactobacillus sp. Marseille-P3519  \n",
       "1926277   Aquibacillus sp. Marseille-P3518  \n",
       "1255658              Luteococcus japonicus  \n",
       "\n",
       "[10575 rows x 8 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d56570c6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:53.478688Z",
     "iopub.status.busy": "2021-10-28T03:58:53.478394Z",
     "iopub.status.idle": "2021-10-28T03:58:53.479921Z",
     "shell.execute_reply": "2021-10-28T03:58:53.479618Z"
    }
   },
   "outputs": [],
   "source": [
    "def concat_slash(x):\n",
    "    #k = x['kingdom']\n",
    "    p = x['phylum']\n",
    "    c = x['class']\n",
    "    o = x['order']\n",
    "    f = x['family']\n",
    "    g = x['genus']\n",
    "    s = x['species']\n",
    "    return '/'.join(list(map(str, [p, c, o, f, g, s])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "24e3331d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:53.481730Z",
     "iopub.status.busy": "2021-10-28T03:58:53.481443Z",
     "iopub.status.idle": "2021-10-28T03:58:53.672219Z",
     "shell.execute_reply": "2021-10-28T03:58:53.672465Z"
    }
   },
   "outputs": [],
   "source": [
    "wol_taxa = list(taxonomy.apply(concat_slash, axis=1).values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3d1403a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:53.674399Z",
     "iopub.status.busy": "2021-10-28T03:58:53.674113Z",
     "iopub.status.idle": "2021-10-28T03:58:54.315940Z",
     "shell.execute_reply": "2021-10-28T03:58:54.315663Z"
    }
   },
   "outputs": [],
   "source": [
    "gpd_metadata = pd.read_table('../results/mmvec/GPD_metadata.tsv')\n",
    "gpd_metadata = gpd_metadata.dropna(subset=['Host_range_isolates'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9f5d6088",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:54.319744Z",
     "iopub.status.busy": "2021-10-28T03:58:54.319378Z",
     "iopub.status.idle": "2021-10-28T03:58:54.556814Z",
     "shell.execute_reply": "2021-10-28T03:58:54.556568Z"
    }
   },
   "outputs": [],
   "source": [
    "lookup = {\n",
    " 'Actinobacteriota' : 'Actinobacteria',\n",
    " 'Bacteroidota' : 'Bacteroidetes',\n",
    " 'Desulfobacterota': 'Deltaproteobacteria',\n",
    " 'Synergistota': 'Synergistetes',\n",
    " 'Campylobacterota': 'Epsilonproteobacteria',\n",
    " 'Fusobacteriota': 'Fusobacteria'}\n",
    "    \n",
    "def fix_f(x):\n",
    "    y = x.split('/')[0]\n",
    "    if y in lookup.keys():\n",
    "        z = lookup[y]\n",
    "        return x.replace(y, z)\n",
    "    return x\n",
    "        \n",
    "def all_taxa(x):\n",
    "    if pd.isnull(x):\n",
    "        return []\n",
    "    else:\n",
    "        taxa = str(x).split(',')\n",
    "        return list(set(list(map(fix_f, taxa))))\n",
    "    \n",
    "gpd_taxa = list(gpd_metadata['Host_range_taxon'].apply(all_taxa).values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f24dd2",
   "metadata": {},
   "source": [
    "# Filter out interactions based on taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "029562e0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:58.362901Z",
     "iopub.status.busy": "2021-10-28T03:58:58.362593Z",
     "iopub.status.idle": "2021-10-28T03:58:58.364639Z",
     "shell.execute_reply": "2021-10-28T03:58:58.364364Z"
    }
   },
   "outputs": [],
   "source": [
    "all_gpd_taxa = sum(gpd_taxa, [])\n",
    "all_gpd_taxa = set(all_gpd_taxa)\n",
    "common_taxa = set(wol_taxa) & set(all_gpd_taxa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2bb35fcc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:58.376098Z",
     "iopub.status.busy": "2021-10-28T03:58:58.370994Z",
     "iopub.status.idle": "2021-10-28T03:58:58.396241Z",
     "shell.execute_reply": "2021-10-28T03:58:58.396484Z"
    }
   },
   "outputs": [],
   "source": [
    "def contains_f(x):\n",
    "    return len(set(x) & common_taxa) > 0\n",
    "\n",
    "idx = list(map(contains_f, gpd_taxa))\n",
    "\n",
    "gpd_subset = gpd_metadata.loc[idx]\n",
    "viruses = gpd_metadata.loc[idx, 'GPD_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "836a0f55",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:58:58.400431Z",
     "iopub.status.busy": "2021-10-28T03:58:58.399854Z",
     "iopub.status.idle": "2021-10-28T03:59:03.467457Z",
     "shell.execute_reply": "2021-10-28T03:59:03.467136Z"
    }
   },
   "outputs": [],
   "source": [
    "averina_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Averina.txt')\n",
    "averina_edges = averina_edges.loc[averina_edges.cond_logit > 1]\n",
    "\n",
    "dan_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Dan.txt')\n",
    "dan_edges = dan_edges.loc[dan_edges.cond_logit > 1]\n",
    "\n",
    "pths_edges = pd.read_table(f'{mmvec_edge_dir}/edges_PTHS.txt')\n",
    "pths_edges = pths_edges.loc[pths_edges.cond_logit > 1]\n",
    "\n",
    "wang_edges = pd.read_table(f'{mmvec_edge_dir}/edges_Wang.txt')\n",
    "wang_edges = wang_edges.loc[wang_edges.cond_logit > 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e1f12b",
   "metadata": {},
   "source": [
    "# Filter out microbes based on taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f231baff",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:03.475099Z",
     "iopub.status.busy": "2021-10-28T03:59:03.470822Z",
     "iopub.status.idle": "2021-10-28T03:59:28.153721Z",
     "shell.execute_reply": "2021-10-28T03:59:28.153436Z"
    }
   },
   "outputs": [],
   "source": [
    "def contains_taxonomy(x):\n",
    "    x = str(x)\n",
    "    if x in taxonomy.index:\n",
    "        tstr = concat_slash(taxonomy.loc[x])\n",
    "        return tstr in common_taxa\n",
    "    return False\n",
    "wang_edges = wang_edges.loc[wang_edges.microbe.apply(contains_taxonomy)]\n",
    "dan_edges = dan_edges.loc[dan_edges.microbe.apply(contains_taxonomy)]\n",
    "averina_edges = averina_edges.loc[averina_edges.microbe.apply(contains_taxonomy)]\n",
    "pths_edges = pths_edges.loc[pths_edges.microbe.apply(contains_taxonomy)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6dd550ab",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:28.156569Z",
     "iopub.status.busy": "2021-10-28T03:59:28.155721Z",
     "iopub.status.idle": "2021-10-28T03:59:28.158787Z",
     "shell.execute_reply": "2021-10-28T03:59:28.158493Z"
    }
   },
   "outputs": [],
   "source": [
    "wang_edges['dataset'] = 'Wang'\n",
    "dan_edges['dataset'] = 'Dan'\n",
    "averina_edges['dataset'] = 'Averina'\n",
    "pths_edges['dataset'] = 'PTHS'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1f030ceb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:28.161293Z",
     "iopub.status.busy": "2021-10-28T03:59:28.161001Z",
     "iopub.status.idle": "2021-10-28T03:59:28.196559Z",
     "shell.execute_reply": "2021-10-28T03:59:28.196296Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-1-37c2e8e6a46e>:2: DeprecationWarning: `np.str` is a deprecated alias for the builtin `str`. To silence this warning, use `str` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.str_` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  all_edges['microbe'] = all_edges['microbe'].astype(np.str)\n"
     ]
    }
   ],
   "source": [
    "all_edges = pd.concat((wang_edges, dan_edges, averina_edges, pths_edges), axis=0)\n",
    "all_edges['microbe'] = all_edges['microbe'].astype(np.str)\n",
    "# merge microbial metadata\n",
    "all_edges = pd.merge(all_edges, gpd_metadata, left_on='virus', right_on='GPD_id')\n",
    "# merge GPD metadata\n",
    "all_edges = pd.merge(all_edges, taxonomy, left_on='microbe', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e9f45cac",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:28.198587Z",
     "iopub.status.busy": "2021-10-28T03:59:28.198299Z",
     "iopub.status.idle": "2021-10-28T03:59:28.479259Z",
     "shell.execute_reply": "2021-10-28T03:59:28.479518Z"
    }
   },
   "outputs": [],
   "source": [
    "all_edges.to_csv('../results/interaction_metadata.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8f9639d1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:28.482357Z",
     "iopub.status.busy": "2021-10-28T03:59:28.482071Z",
     "iopub.status.idle": "2021-10-28T03:59:28.502011Z",
     "shell.execute_reply": "2021-10-28T03:59:28.502260Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>microbe</th>\n",
       "      <th>virus</th>\n",
       "      <th>cond_logit</th>\n",
       "      <th>dataset</th>\n",
       "      <th>GPD_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>GPD_VC</th>\n",
       "      <th>Size</th>\n",
       "      <th>Predicted_phage_taxon</th>\n",
       "      <th>Host_range_isolates</th>\n",
       "      <th>...</th>\n",
       "      <th>checkV_termini</th>\n",
       "      <th>Novel</th>\n",
       "      <th>genome</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29466</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>4.949254</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18797</td>\n",
       "      <td>40415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000011045</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G001553335</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>Veillonella parvula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>29466</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>4.949254</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2376</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18797</td>\n",
       "      <td>40415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000011045</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G001553335</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>Veillonella parvula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>29466</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>4.537919</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>41768</td>\n",
       "      <td>39592</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003602765</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G001553335</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>Veillonella parvula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>29466</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>4.537919</td>\n",
       "      <td>Wang</td>\n",
       "      <td>ivig_2256</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>41768</td>\n",
       "      <td>39592</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003602765</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G001553335</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>Veillonella parvula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>29466</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>4.431313</td>\n",
       "      <td>Wang</td>\n",
       "      <td>uvig_108002</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>37815</td>\n",
       "      <td>32601</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>GCA_003464755,GCF_000164195,12718_7#19,20298_3...</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G001553335</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>Veillonella parvula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3198</th>\n",
       "      <td>1871003</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>1.122697</td>\n",
       "      <td>Averina</td>\n",
       "      <td>ivig_4295</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>18307</td>\n",
       "      <td>23261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21673_4#3</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3238</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_173554</td>\n",
       "      <td>1.054612</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_173554</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>3999</td>\n",
       "      <td>60158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003482185</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3242</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_173554</td>\n",
       "      <td>1.054612</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_173554</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>3999</td>\n",
       "      <td>60158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003482185</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>1.052579</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>3999</td>\n",
       "      <td>59704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003482185</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3250</th>\n",
       "      <td>1871003</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>1.052579</td>\n",
       "      <td>Averina</td>\n",
       "      <td>uvig_492171</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>3999</td>\n",
       "      <td>59704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003482185</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>G900104605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Tidjanibacter</td>\n",
       "      <td>Tidjanibacter massiliensis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3356 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      microbe        virus  cond_logit  dataset       GPD_id      Source  \\\n",
       "0       29466    ivig_2376    4.949254     Wang    ivig_2376     Isolate   \n",
       "8       29466    ivig_2376    4.949254     Wang    ivig_2376     Isolate   \n",
       "16      29466    ivig_2256    4.537919     Wang    ivig_2256     Isolate   \n",
       "24      29466    ivig_2256    4.537919     Wang    ivig_2256     Isolate   \n",
       "32      29466  uvig_108002    4.431313     Wang  uvig_108002  Metagenome   \n",
       "...       ...          ...         ...      ...          ...         ...   \n",
       "3198  1871003    ivig_4295    1.122697  Averina    ivig_4295     Isolate   \n",
       "3238  1871003  uvig_173554    1.054612  Averina  uvig_173554  Metagenome   \n",
       "3242  1871003  uvig_173554    1.054612  Averina  uvig_173554  Metagenome   \n",
       "3246  1871003  uvig_492171    1.052579  Averina  uvig_492171  Metagenome   \n",
       "3250  1871003  uvig_492171    1.052579  Averina  uvig_492171  Metagenome   \n",
       "\n",
       "      GPD_VC   Size Predicted_phage_taxon  \\\n",
       "0      18797  40415                   NaN   \n",
       "8      18797  40415                   NaN   \n",
       "16     41768  39592                   NaN   \n",
       "24     41768  39592                   NaN   \n",
       "32     37815  32601            Myoviridae   \n",
       "...      ...    ...                   ...   \n",
       "3198   18307  23261                   NaN   \n",
       "3238    3999  60158                   NaN   \n",
       "3242    3999  60158                   NaN   \n",
       "3246    3999  59704                   NaN   \n",
       "3250    3999  59704                   NaN   \n",
       "\n",
       "                                    Host_range_isolates  ... checkV_termini  \\\n",
       "0                                         GCF_000011045  ...             No   \n",
       "8                                         GCF_000011045  ...             No   \n",
       "16                                        GCA_003602765  ...             No   \n",
       "24                                        GCA_003602765  ...             No   \n",
       "32    GCA_003464755,GCF_000164195,12718_7#19,20298_3...  ...             No   \n",
       "...                                                 ...  ...            ...   \n",
       "3198                                          21673_4#3  ...             No   \n",
       "3238                                      GCA_003482185  ...             No   \n",
       "3242                                      GCA_003482185  ...             No   \n",
       "3246                                      GCA_003482185  ...             No   \n",
       "3250                                      GCA_003482185  ...             No   \n",
       "\n",
       "     Novel      genome   kingdom         phylum          class  \\\n",
       "0      Yes  G001553335  Bacteria     Firmicutes  Negativicutes   \n",
       "8      Yes  G001553335  Bacteria     Firmicutes  Negativicutes   \n",
       "16     Yes  G001553335  Bacteria     Firmicutes  Negativicutes   \n",
       "24     Yes  G001553335  Bacteria     Firmicutes  Negativicutes   \n",
       "32     Yes  G001553335  Bacteria     Firmicutes  Negativicutes   \n",
       "...    ...         ...       ...            ...            ...   \n",
       "3198    No  G900104605  Bacteria  Bacteroidetes    Bacteroidia   \n",
       "3238   Yes  G900104605  Bacteria  Bacteroidetes    Bacteroidia   \n",
       "3242   Yes  G900104605  Bacteria  Bacteroidetes    Bacteroidia   \n",
       "3246   Yes  G900104605  Bacteria  Bacteroidetes    Bacteroidia   \n",
       "3250   Yes  G900104605  Bacteria  Bacteroidetes    Bacteroidia   \n",
       "\n",
       "               order           family          genus  \\\n",
       "0     Veillonellales  Veillonellaceae    Veillonella   \n",
       "8     Veillonellales  Veillonellaceae    Veillonella   \n",
       "16    Veillonellales  Veillonellaceae    Veillonella   \n",
       "24    Veillonellales  Veillonellaceae    Veillonella   \n",
       "32    Veillonellales  Veillonellaceae    Veillonella   \n",
       "...              ...              ...            ...   \n",
       "3198   Bacteroidales    Rikenellaceae  Tidjanibacter   \n",
       "3238   Bacteroidales    Rikenellaceae  Tidjanibacter   \n",
       "3242   Bacteroidales    Rikenellaceae  Tidjanibacter   \n",
       "3246   Bacteroidales    Rikenellaceae  Tidjanibacter   \n",
       "3250   Bacteroidales    Rikenellaceae  Tidjanibacter   \n",
       "\n",
       "                         species  \n",
       "0            Veillonella parvula  \n",
       "8            Veillonella parvula  \n",
       "16           Veillonella parvula  \n",
       "24           Veillonella parvula  \n",
       "32           Veillonella parvula  \n",
       "...                          ...  \n",
       "3198  Tidjanibacter massiliensis  \n",
       "3238  Tidjanibacter massiliensis  \n",
       "3242  Tidjanibacter massiliensis  \n",
       "3246  Tidjanibacter massiliensis  \n",
       "3250  Tidjanibacter massiliensis  \n",
       "\n",
       "[3356 rows x 29 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "caa7692a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a01f1749",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T03:59:28.510473Z",
     "iopub.status.busy": "2021-10-28T03:59:28.510187Z",
     "iopub.status.idle": "2021-10-28T03:59:28.512126Z",
     "shell.execute_reply": "2021-10-28T03:59:28.511824Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>398511</th>\n",
       "      <td>G000005825</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillus pseudofirmus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456320</th>\n",
       "      <td>G000006175</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanococcaceae</td>\n",
       "      <td>Methanococcus</td>\n",
       "      <td>Methanococcus voltae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306537</th>\n",
       "      <td>G000006605</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Corynebacterium</td>\n",
       "      <td>Corynebacterium jeikeium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160492</th>\n",
       "      <td>G000006725</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Xylella</td>\n",
       "      <td>Xylella fastidiosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243277</th>\n",
       "      <td>G000006745</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484498</th>\n",
       "      <td>G900156675</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillaceae</td>\n",
       "      <td>Oleibacter</td>\n",
       "      <td>Oleibacter marinus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477680</th>\n",
       "      <td>G900156765</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Chitinophagia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Filimonas</td>\n",
       "      <td>Filimonas lacunae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926284</th>\n",
       "      <td>G900156885</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillus sp. Marseille-P3519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926277</th>\n",
       "      <td>G900156915</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Aquibacillus</td>\n",
       "      <td>Aquibacillus sp. Marseille-P3518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255658</th>\n",
       "      <td>G900163845</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Propionibacteriales</td>\n",
       "      <td>Propionibacteriaceae</td>\n",
       "      <td>Luteococcus</td>\n",
       "      <td>Luteococcus japonicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10575 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             genome   kingdom          phylum                class  \\\n",
       "1                                                                    \n",
       "398511   G000005825  Bacteria      Firmicutes              Bacilli   \n",
       "456320   G000006175   Archaea   Euryarchaeota         Methanococci   \n",
       "306537   G000006605  Bacteria  Actinobacteria       Actinobacteria   \n",
       "160492   G000006725  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "243277   G000006745  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "...             ...       ...             ...                  ...   \n",
       "484498   G900156675  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "477680   G900156765  Bacteria   Bacteroidetes        Chitinophagia   \n",
       "1926284  G900156885  Bacteria      Firmicutes              Bacilli   \n",
       "1926277  G900156915  Bacteria      Firmicutes              Bacilli   \n",
       "1255658  G900163845  Bacteria  Actinobacteria       Actinobacteria   \n",
       "\n",
       "                       order                family            genus  \\\n",
       "1                                                                     \n",
       "398511            Bacillales           Bacillaceae         Bacillus   \n",
       "456320       Methanococcales      Methanococcaceae    Methanococcus   \n",
       "306537     Corynebacteriales    Corynebacteriaceae  Corynebacterium   \n",
       "160492       Xanthomonadales      Xanthomonadaceae          Xylella   \n",
       "243277           Vibrionales          Vibrionaceae           Vibrio   \n",
       "...                      ...                   ...              ...   \n",
       "484498     Oceanospirillales    Oceanospirillaceae       Oleibacter   \n",
       "477680       Chitinophagales      Chitinophagaceae        Filimonas   \n",
       "1926284      Lactobacillales      Lactobacillaceae    Lactobacillus   \n",
       "1926277           Bacillales           Bacillaceae     Aquibacillus   \n",
       "1255658  Propionibacteriales  Propionibacteriaceae      Luteococcus   \n",
       "\n",
       "                                   species  \n",
       "1                                           \n",
       "398511               Bacillus pseudofirmus  \n",
       "456320                Methanococcus voltae  \n",
       "306537            Corynebacterium jeikeium  \n",
       "160492                  Xylella fastidiosa  \n",
       "243277                     Vibrio cholerae  \n",
       "...                                    ...  \n",
       "484498                  Oleibacter marinus  \n",
       "477680                   Filimonas lacunae  \n",
       "1926284  Lactobacillus sp. Marseille-P3519  \n",
       "1926277   Aquibacillus sp. Marseille-P3518  \n",
       "1255658              Luteococcus japonicus  \n",
       "\n",
       "[10575 rows x 8 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f39691",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
